### PR TITLE
Cleanups in some hp::DoFHandler functions.

### LINKED ISSUE #7162: Bug in hp::DoFHandler::distribute_dofs()
@Rombur had better eyes than I in https://github.com/dealii/dealii/pull/7150#pullrequestreview-153096688 and found a bug in the current hp parallel implementation (see #3511).

@marcfehling: Here is a testcase that unfortunately requires a 16x16 grid, Q2 elements, and at least 9 processors:
```
// ---------------------------------------------------------------------
//
// Copyright (C) 2018 by the deal.II authors
//
// This file is part of the deal.II library.
//
// The deal.II library is free software; you can use it, redistribute
// it, and/or modify it under the terms of the GNU Lesser General
// Public License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
// The full text of the license can be found in the file LICENSE.md at
// the top level directory of deal.II.
//
// ---------------------------------------------------------------------


// A vastly reduced derivative of step-40 that uses hp finite elements
// and makes sure that the number of degrees of freedom is independent
// of the number of processors. It does so by creating subsets of the
// MPI universe of different number of processes, and storing the
// number of unknowns on each.

#include <deal.II/base/utilities.h>

#include <deal.II/distributed/tria.h>

#include <deal.II/dofs/dof_accessor.h>

#include <deal.II/fe/fe_q.h>

#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/tria_accessor.h>
#include <deal.II/grid/tria_iterator.h>

#include <deal.II/hp/dof_handler.h>
#include <deal.II/hp/fe_collection.h>

#include <iostream>

#include "../tests.h"



template <int dim>
void
test(MPI_Comm mpi_communicator)
{
  parallel::distributed::Triangulation<dim> triangulation(mpi_communicator);

  hp::FECollection<dim> fe;
  fe.push_back(FE_Q<dim>(2));
  fe.push_back(FE_Q<dim>(2));
  fe.push_back(FE_Q<dim>(2));

  GridGenerator::hyper_cube(triangulation);
  triangulation.refine_global(4);

  hp::DoFHandler<dim> dof_handler(triangulation);

  // set active_fe_index mostly randomly
  for (const auto &cell : dof_handler.active_cell_iterators())
    if (cell->is_locally_owned())
      cell->set_active_fe_index(cell->active_cell_index() % fe.size());

  dof_handler.distribute_dofs(fe);

  if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
    deallog << "   n_procs/n_dofs: "
            << Utilities::MPI::n_mpi_processes(mpi_communicator) << '/'
            << dof_handler.n_dofs() << std::endl;
}



int
main(int argc, char *argv[])
{
  Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);
  MPILogInitAll                    log;

  // Create communicators with larger and larger subsets of
  // processors, and then see what happens with the number of DoFs we
  // get on the (always same) mesh created in the test() function
  for (unsigned int n = 1; n <= Utilities::MPI::n_mpi_processes(MPI_COMM_WORLD);
       ++n)
    {
      // Colorize the current processor: zero if we want to include it
      // in the communication, one otherwise
      const int color =
        (Utilities::MPI::this_mpi_process(MPI_COMM_WORLD) < n ? 0 : 1);

      // Then split the communicator into two, based on the two
      // colors. We'll only do something with that communicator that
      MPI_Comm subset_comm;
      MPI_Comm_split(MPI_COMM_WORLD, color, /*key=*/0, &subset_comm);

      if (color == 0)
        test<2>(subset_comm);

      // destroy the communicators previously created
      MPI_Comm_free(&subset_comm);
    }
}
```
It runs through 1-N processes and outputs, for the ever same mesh, how many DoFs we have distributed. I currently get this when run with `mpirun -np 10`:
```
1: ! DEAL:0::   n_procs/n_dofs: 1/1089
1: ! DEAL:0::   n_procs/n_dofs: 2/1089
1: ! DEAL:0::   n_procs/n_dofs: 3/1089
1: ! DEAL:0::   n_procs/n_dofs: 4/1089
1: ! DEAL:0::   n_procs/n_dofs: 5/1089
1: ! DEAL:0::   n_procs/n_dofs: 6/1089
1: ! DEAL:0::   n_procs/n_dofs: 7/1089
1: ! DEAL:0::   n_procs/n_dofs: 8/1089
1: ! DEAL:0::   n_procs/n_dofs: 9/1093
1: ! DEAL:0::   n_procs/n_dofs: 10/1093
```
Note how we start with 1089 DoFs (correct), but then switch to  1093 (wrong). The algorithm should yield a number of DoFs independent of the partitioning.

### PR DESCRIPTION
**Pull request by bangerth**
This is in preparation for a patch to resolve #7162 after
@marcfehling explained the bug to me. That patch is coming
next but I thought I'd get the clean-ups out of the way
already given that these cleanups significantly obscure
what the real patch is going to be.

In pursuance of #3511.

### CODE REVIEW

**Comment by bangerth** [comment]
Would appreciate a quick turn-around on this one.

File: source/hp/dof_handler.cc [hunk 1]
```diff
@@ -464,19 +463,19 @@ namespace internal
...
-            // now that we know how many face dofs we will have to
-            // have on each level, allocate the memory. note that we
+            // Now that we know how many face dofs we will have to
+            // have, allocate the memory. Note that we
             // allocate offsets for all faces, though only the active
             // ones will have a non-invalid value later on
             face_dof_offsets =
               std::vector<unsigned int>(dof_handler.tria->n_raw_faces(),
-                                        (unsigned int)(-1));
+                                        static_cast<unsigned int>(-1));
```
Comment (drwells) [inline source/hp/dof_handler.cc]: Would it be better to use `numbers::invalid_unsigned_int` here?

File: source/hp/dof_handler.cc [hunk 2]
```diff
@@ -439,15 +438,15 @@ namespace internal
...
                           1 +
                           dof_handler.get_fe(cell->active_fe_index())
                             .template n_dofs_per_object<dim - 1>() +
                           1;
 
-                      // otherwise we do indeed need two sets, i.e. two
+                      // Otherwise we do indeed need two sets, i.e. two
                       // active_fe_indices, two sets of dofs, and one stop
                       // index:
                       else
```
Comment (drwells) [inline source/hp/dof_handler.cc]: Would you mind adding the same comments here?

Comment (drwells) [reply source/hp/dof_handler.cc]: (that is, lines 454-460)

File: source/hp/dof_handler.cc [hunk 3]
```diff
@@ -464,19 +463,19 @@ namespace internal
...
 
-            // now that we know how many face dofs we will have to
-            // have on each level, allocate the memory. note that we
+            // Now that we know how many face dofs we will have to
+            // have, allocate the memory. Note that we
             // allocate offsets for all faces, though only the active
             // ones will have a non-invalid value later on
             face_dof_offsets =
               std::vector<unsigned int>(dof_handler.tria->n_raw_faces(),
-                                        (unsigned int)(-1));
```
Comment (davydden) [inline source/hp/dof_handler.cc]: Should not we use numbers::invalid_unsigned_int?

**Comment by drwells** [comment]
Also, if you are feeling fancy, we could use
```
for (const auto &face : triangulation.active_face_iterators()
```

instead of using user flags.

**Comment by bangerth** [comment]
On 09/26/2018 03:31 PM, David Wells wrote:
> Also, if you are feeling fancy, we could use
> 
> |for (const auto &face : triangulation.active_face_iterators() |
> 
> instead of using user flags.

Nope, I can't do that because I need to get the two adjacent cells to 
find out what elements live there. There is no way to go from faces to 
cells :-(

**Comment by drwells** [comment]
Aw rats, you're right.

**Comment by bangerth** [comment]
I made the requested changes and also created a small array to hold the two `active_fe_index` values in the second block that I edited. (You will see in the follow-up why.)

This passes all `.*hp.*` tests on my system.

File: source/hp/dof_handler.cc [hunk 4]
```diff
@@ -525,61 +524,74 @@ namespace internal
...
+                          // There are two active_fe_indices that live on this
+                          // face.
                           face_dof_offsets[cell->face(face)->index()] =
                             next_free_face_slot;
 
-                          // set first slot for this face to
+                          // Store the two indices we will have to deal with.
+                          unsigned int active_fe_indices[2] = {
+                            cell->active_fe_index(),
+                            cell->neighbor(face)->active_fe_index()};
```
Comment (drwells) [inline source/hp/dof_handler.cc]: :+1:

**Pull request merged**